File: mirror/tier2-php/showcase/meta_decorators.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Metaprogramming
// Module: meta_decorators.php

function timing_decorator(callable $func): callable {
    return function(...$args) use ($func) {
        $start = microtime(true);
        $result = $func(...$args);
        $elapsed = microtime(true) - $start;
        return $result;
    };
}

function memoization_decorator(callable $func): callable {
    $cache = [];
    return function(...$args) use ($func, &$cache) {
        $key = json_encode($args);
        if (!isset($cache[$key])) {
            $cache[$key] = $func(...$args);
        }
        return $cache[$key];
    };
}

function validation_decorator(callable $validator, callable $func): callable {
    return function(...$args) use ($validator, $func) {
        if (!$validator(...$args)) return null;
        return $func(...$args);
    };
}

$fibonacci = memoization_decorator(function($n) use (&$fibonacci) {
    return $n < 2 ? $n : $fibonacci($n - 1) + $fibonacci($n - 2);
});

$result = [
    'memo_fib_5' => $fibonacci(5),
    'memo_fib_10' => $fibonacci(10),
    'decorators_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-ruby/showcase/meta_decorators.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Metaprogramming
# Module: meta_decorators.rb
# Purpose: Demonstrate decorator patterns

def timing_decorator(&block)
  -> (*args) do
    start = Time.now
    result = block.call(*args)
    elapsed = Time.now - start
    result
  end
end

def memoization_decorator(&block)
  cache = {}
  -> (*args) do
    if cache.key?(args)
      cache[args]
    else
      cache[args] = block.call(*args)
    end
  end
end

def validation_decorator(validator, &block)
  -> (*args) do
    return nil unless validator.call(*args)
    block.call(*args)
  end
end

fibonacci = memoization_decorator do |n|
  n < 2 ? n : fibonacci.call(n-1) + fibonacci.call(n-2)
end

result = {
  memo_fib_5: fibonacci.call(5),
  memo_fib_10: fibonacci.call(10),
  decorators_active: true,
}

puts result.inspect
